plugins/field-time/src/field_time.js

```diff
@@ -0,0 +1,115 @@
+/**
+ * @license
+ * Copyright 2020 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @fileoverview Time input field.
+ * @author carl.shen1@uwaterloo.ca (Carl Shen)
+ */
+
+import * as Blockly from 'blockly/core';
+
+/**
+ * Class for a time input field.
+ * @extends {Blockly.FieldTextInput}
+ */
+export class FieldTime extends Blockly.FieldTextInput {
+  /**
+   * Class for a time input field. Built using <input type='time'/> time picker
+   * @param {string=} value The initial value of the field. Should be in
+   *    'HH:MM' format. Defaults to 12:00.
+   * @param {Function=} validator A function that is called to validate
+   *    changes to the field's value. Takes in a time string & returns a
+   *    validated time string ('HH:MM' format), or null to abort the
+   *    change.
+   */
+  constructor(value = '12:00', validator = undefined) {
+    super('12:00', validator);
+    this.setEditorValue_(value);
+    this.CURSOR = 'pointer';
+    this.timeInput = null;
+  }
+
+  /**
+   * Constructs a FieldTime from a JSON arg object.
+   * @param {!Object} options A JSON object with options (time).
+   * @return {!FieldTime} The new field instance.
+   * @package
+   * @nocollapse
+   */
+  static fromJson(options) {
+    return new FieldTime(options['time'], undefined);
+  }
+
+  /**
+   * Ensures that the input value is a valid time.
+   * @param {*=} newValue The input value.
+   * @return {?string} A valid time, or null if invalid.
+   * @protected
+   */
+  doClassValidation_(newValue = undefined) {
+    if (!newValue) {
+      return null;
+    }
+    const isValid =
+      /^([0-1]?[0-9]|2[0-4]):([0-5][0-9])(:[0-5][0-9])?$/.test(newValue);
```

Comment:
* Do you want to accept `24:00`?  Do you want to accept `24:59`?
* Do you want to accept seconds?  The description suggests it's strictly HH:MM.
* The middle set of parens appear to be superfluous:
  ```suggestion
        /^([0-1]?[0-9]|2[0-4]):[0-5][0-9](:[0-5][0-9])?$/.test(newValue);
  ```
* An alternative approach might be to match against `/^(\d\d?):(\d\d)(?::(\d\d))?$/` and then check the submatches numerically.  (Probably not simpler, but worth keeping in mind if the regexp starts getting out of hand.)